### PR TITLE
Add a generic SPI NOR flash driver instead of the specific w25qxxdv driver

### PR DESCRIPTION
**Pull request by sbourdelin**
Implement a new generic SPI NOR flash driver using the CFI (Common Flash Interface) and the JEDEC ID to address all SPI NOR flash.
This driver is heavily based on the previous w25qxxdv, with few enhancements and a flash id table which contains the geometry for few SPI NOR flash.

### CODE REVIEW

**Comment by sbourdelin** [comment]
@nashif ping?

**Comment by vanwinkeljan** [comment]
Looks to me that there is already more generic SPI NOR driver pull request #7561 (which is based on pr #5456). PR #7561 relies on the SFDP information present in the flash device instead of adding the data hard coded into a struct for every device.

Is till see a use case for this driver if the parameters where passed from dts or kconfig as this would reduce the code size compared to auto detection at run time.

File: drivers/flash/spi_nor.c [hunk 1]
```diff
@@ -0,0 +1,510 @@
...
+static const struct flash_info {
+	char *name;
+	u8_t id[SPI_NOR_MAX_ID_LEN];
+	unsigned page_size;
+	unsigned sector_size;
+	unsigned n_sectors;
+	unsigned block_size;
+	uint32_t flag;
+} flash_ids[] = {
+	{ "W25Q80",     JEDEC_ID(0xef4014), SZ_256, SZ_4K, SZ_256, SZ_64K, BLOCK_ERASE_32K | BLOCK_ERASE_64K },
```
Comment (tbursztyka) [inline drivers/flash/spi_nor.c]: 80 chars limit (why checkpatch has not noticed it?)

File: drivers/flash/spi_nor.c [hunk 2]
```diff
@@ -0,0 +1,510 @@
...
+	}
+
+err:
+	SYNC_UNLOCK();
+
+	return (dest - p_dest);
+}
+
+/**
+ * @brief Write data to the flash
```
Comment (tbursztyka) [inline drivers/flash/spi_nor.c]: You don't need to repeat the documentation which is the same as flash API. So remove it for write, read, etc...

File: drivers/flash/spi_nor.c [hunk 3]
```diff
@@ -0,0 +1,510 @@
...
+#define SYNC_INIT() k_sem_init(	\
+		&((struct spi_nor_data *)dev->driver_data)->sem, 1, UINT_MAX)
+#define SYNC_LOCK() k_sem_take(&driver_data->sem, K_FOREVER)
+#define SYNC_UNLOCK() k_sem_give(&driver_data->sem)
+#else
+#define SYNC_INIT()
+#define SYNC_LOCK()
+#define SYNC_UNLOCK()
+#endif
+
```
Comment (tbursztyka) [inline drivers/flash/spi_nor.c]: move all the above to the spi_nor.h header file

File: drivers/flash/spi_nor.c [hunk 4]
```diff
@@ -0,0 +1,510 @@
...
+		.count = 2
+	};
+
+	if (is_write) {
+		return spi_write(driver_data->spi, &driver_data->spi_cfg, &tx_set);
+	}
+
+	return spi_transceive(driver_data->spi, &driver_data->spi_cfg, &tx_set, &rx_set);
+}
+#define spi_nor_cmd_read(dev, opcode, dest, length) \
```
Comment (tbursztyka) [inline drivers/flash/spi_nor.c]: add an empty line here. I'd rather use inlines, but that's fine

File: drivers/flash/spi_nor.c [hunk 5]
```diff
@@ -0,0 +1,510 @@
...
+	return 0;
+}
+
+/**
+ * @brief Wait until the flash is ready
+ *
+ * @param dev The device structure
+ * @return 0 on success, negative errno code otherwise
+ */
+static inline int spi_nor_wait_until_ready(struct device *dev)
```
Comment (tbursztyka) [inline drivers/flash/spi_nor.c]: no need to make it inline I think

**Review by tbursztyka — CHANGES_REQUESTED** [review]
As noticed in #10988, having a patch that enables dts would be nice and avoid the Kconfig option about SPI settings.

**Comment by nashif** [comment]
merged in another PR #10988,

**Comment by sbourdelin** [comment]
Sorry for the delay.
Thanks for the review and the merge in the PR #10988

Regards,
Sébastien.

**Pull request closed without merging**